src/CiviEntityStorage.php

```diff
@@ -0,0 +1,181 @@
+<?php
+// In construct make sure to invoke initialize
+namespace Drupal\civicrm_entity;
+
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\Core\Entity\ContentEntityStorageBase;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Entity\EntityManagerInterface;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Field\FieldDefinitionInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Defines entity class for external CiviCRM entities.
+ */
+class CiviEntityStorage extends ContentEntityStorageBase {
+
+  /**
+   * @var \Drupal\civicrm_entity\CiviCrmApi
+   */
+  protected $civicrmApi;
+
+  public function __construct(EntityTypeInterface $entity_type, EntityManagerInterface $entity_manager, CacheBackendInterface $cache, CiviCrmApi $civicrm_api) {
+    parent::__construct($entity_type, $entity_manager, $cache);
+    $this->civicrmApi = $civicrm_api;
+  }
+
+  public static function createInstance(ContainerInterface $container, EntityTypeInterface $entity_type) {
+    return new static(
+      $entity_type,
+      $container->get('entity.manager'),
+      $container->get('cache.entity'),
+      $container->get('civicrm_entity.api')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function doDelete($entities) {
+    /** @var EntityInterface $entity */
+    foreach ($entities as $entity) {
+      try {
+        $params['id'] = $entity->id();
+        $this->civicrmApi->delete($this->entityType->get('civicrm_entity'), $params);
+      }
+      catch (\Exception $e) {
+        throw $e;
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function doSave($id, EntityInterface $entity) {
+    if ($entity->isNew()) {
+      // @todo decide on special handling?
+      $return = SAVED_NEW;
+    }
+    else {
+      $return = SAVED_UPDATED;
+    }
+
+    $params = $entity->toArray();
+    $params = array_map(function ($value) {
```

Comment:
It'd be better if this code was more explicit, ie. looked at field values and decomposed them based on their expected structure. We should be able to know how a field will be represented in advance (based on how it was defined), so if it's going to be `[ 'und' => [ 0 => [ 'value' ] ] ]`, let's explicitly take that apart.